contracts/marketplace/src/contract.rs

```diff
@@ -0,0 +1,444 @@
+    info: MessageInfo,
+    listing_id: String,
+) -> Result<Response, ContractError> {
+    let listing = listings().load(deps.storage, listing_id.clone())?;
+    ensure_eq!(
+        listing.seller,
+        info.sender,
+        ContractError::Unauthorized {
+            message: "sender is not the seller".to_string(),
+        }
+    );
+    // can't cancel a list that is pending approval if sale approvals are enabled
+    // listings that are in pending status have already been placed a matching buy order
+    // but it's not yet been accepted by the manager
+    if CONFIG.load(deps.storage)?.sale_approvals && listing.status != ListingStatus::Active {
+        return Err(ContractError::InvalidListingStatus {
+            expected: ListingStatus::Active.to_string(),
+            actual: listing.status.to_string(),
+        });
+    }
+
+    listings().remove(deps.storage, listing_id.clone())?;
+    // query if there is a listing in the asset contract (in case is out of sync)
+    let asset_listing = query_listing(&deps.querier, &listing.collection, &listing.token_id);
+
+    let mut sub_msgs = vec![];
+
+    if asset_listing.is_ok() {
+        let cancel_listing = asset::msg::ExecuteMsg::<
+            cw721::DefaultOptionalNftExtensionMsg,
+            cw721::DefaultOptionalCollectionExtensionMsg,
+            asset::msg::AssetExtensionExecuteMsg,
+        >::UpdateExtension {
+            msg: asset::msg::AssetExtensionExecuteMsg::Delist {
+                token_id: listing.token_id.clone(),
+            },
+        };
+        sub_msgs.push(WasmMsg::Execute {
+            contract_addr: listing.collection.to_string(),
+            msg: to_json_binary(&cancel_listing)?,
+            funds: vec![],
+        });
+    }
+    Ok(Response::new()
+        .add_event(cancel_listing_event(
+            listing_id,
+            listing.collection.clone(),
+            listing.seller,
+            listing.token_id,
+        ))
+        .add_messages(sub_msgs))
+}
+
+pub fn execute_buy_item(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    listing_id: String,
+    price: Coin,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    let listing = listings().load(deps.storage, listing_id.clone())?;
+
+    // Prevent price mismatch due to possible frontrunning
+    if listing.price != price {
+        return Err(ContractError::InvalidPrice {
+            expected: listing.price,
+            actual: price,
+        });
+    }
+
+    // Check payment and funds are valid
+    valid_payment(&info, price.clone(), listing.price.denom.clone())?;
+
+    // if approvals are enabled, create pending sale.
+    if config.sale_approvals {
+        return execute_create_pending_sale(deps, env, info, listing_id, listing, price);
+    }
+
+    // remove listing
+    listings().remove(deps.storage, listing_id.clone())?;
+
+    let buy_msg = asset_buy_msg(info.sender.clone(), listing.token_id.clone());
```

Comment:
As @emperorjm mentioned, this requires the buyer to be the msg sender. This will not work for things like crossmint, gift purchasing, etc. Like the buy handler for the asset, we should accept an optional receiver, and if it is unset, use the info.sender

Reply:
Also, this logic seems to have no enforcement of marketplace fees. How do the marketplace and asset work together for fees?